### PR TITLE
[#11120] [#11121] Implement HTTP request interceptor and structured logging

### LINKED ISSUE #11120: Implement own timer to prevent instance being terminated due to timeout
In GAE Java 11 runtime, there is no `DeadlineExceededException` anymore. Instead, if the request exceeds the 10-min deadline, the entire VM instance whereby the request originates from is terminated. We do not want this to happen and thus we have to implement our own timer to track the request time.

### LINKED ISSUE #11121: Group logs of the same HTTP request together using trace ID
In GAE Java 8 runtime, all logs under the same HTTP request are automatically grouped together.

In GAE Java 11 runtime, there is no such automatic grouping anymore and it makes the logs hard to follow. We need to manually add back the grouping logic by making use of trace ID, as suggested in <https://cloud.google.com/appengine/docs/standard/java11/writing-application-logs#writing_structured_logs>.

### PR DESCRIPTION
**Pull request by wkurniawan07**
Fixes #11120 
Fixes #11121 

Fixing two issues together because both involves creating a filter to intercept incoming HTTP requests.

The last commit enables [structured logging](https://cloud.google.com/logging/docs/structured-logging).

### CODE REVIEW

File: src/main/java/teammates/common/util/Logger.java [hunk 1]
```diff
@@ -18,38 +28,98 @@ public static Logger getLogger() {
...
+            if (traceAndSpan.length == 2) {
+                payload.put("logging.googleapis.com/spanId", traceAndSpan[1].split(";")[0]);
+            }
+        }
+
+        return payload;
+    }
+
+    private StackTraceElement getLoggerSource() {
+        StackTraceElement[] stes = Thread.currentThread().getStackTrace();
```
Comment (rrtheonlyone) [inline src/main/java/teammates/common/util/Logger.java:112]: Getting a `stackTrace()` for every log we do and sweeping through it for the first element is a performance hit. `stackTrace()` walks through the stack and composes an array from it, even if we only need one element. 

Instead of that we can use [StackWalker](https://docs.oracle.com/javase/9/docs/api/java/lang/StackWalker.html). It is also thread safe and is more optimized - it does not create huge memory structures that go unused.

I do not have metrics to back this up but I am pretty sure this is an improvement here.

File: src/main/java/teammates/common/util/Logger.java [hunk 2]
```diff
@@ -18,38 +28,98 @@ public static Logger getLogger() {
...
+        for (int i = 0; i < stes.length; i++) {
+            StackTraceElement ste = stes[i];
+            if (ste.getClassName().equals(Logger.class.getName()) && i + 1 < stes.length) {
+                StackTraceElement nextSte = stes[i + 1];
+                if (!nextSte.getClassName().equals(Logger.class.getName())) {
+                    return nextSte;
+                }
+            }
+        }
+        return null;
```
Comment (rrtheonlyone) [inline src/main/java/teammates/common/util/Logger.java:122]: We should not be calling this when we are at the bottom of the stack ideally - seems odd that we have to check for `null` everywhere from the return of this method.

File: src/main/java/teammates/ui/webapi/RequestTraceFilter.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        // Reference: https://cloud.google.com/appengine/docs/standard/java/taskqueue/push/creating-handlers#reading_request_headers
+        boolean isRequestFromAppEngineQueue = request.getHeader("X-AppEngine-QueueName") != null;
+
+        // GAE will terminate an instance if any request exceeds 10 minutes.
+        // For GAE-invoked requests, we set the limit here minus a small grace period of 5 seconds
+        // to ensure that the 10 minutes limit will not be exceeded.
+        // For user-invoked requests, we keep the time limit at 1 minute (as how it was
+        // in the previous GAE runtime environment) in order to not let user wait for excessively long,
+        // as well as a reminder for us to keep optimizing our API response time.
+        int timeoutInSeconds = isRequestFromAppEngineQueue ? 10 * 60 - 5 : 60;
```
Comment (rrtheonlyone) [inline src/main/java/teammates/ui/webapi/RequestTraceFilter.java:52]: Is the grace period of `5 seconds` sufficient? The instance might still get terminated while we do the teardown steps.

File: src/main/java/teammates/ui/webapi/OriginCheckFilter.java [hunk 4]
```diff
@@ -71,6 +72,16 @@ public void doFilter(ServletRequest req, ServletResponse res, FilterChain chain)
             return;
         }
 
+        // The header X-AppEngine-QueueName cannot be spoofed as GAE will strip any user-sent X-AppEngine-QueueName headers.
+        // Reference: https://cloud.google.com/appengine/docs/standard/java/taskqueue/push/creating-handlers#reading_request_headers
+        boolean isRequestFromAppEngineQueue = request.getHeader("X-AppEngine-QueueName") != null;
+
+        if (isRequestFromAppEngineQueue) {
```
Comment (rrtheonlyone) [inline src/main/java/teammates/ui/webapi/OriginCheckFilter.java:79]: curious on why you added this logic here, I would have thought this was sufficient:
```
    <filter-mapping>
        <filter-name>OriginCheckFilter</filter-name>
        <url-pattern>/webapi/*</url-pattern>
    </filter-mapping
```

Since the `filter` will never be called for those `worker/` or `auto/` App Engine Queue tasks we have.

**Review by rrtheonlyone — CHANGES_REQUESTED** [review]
Thanks for the work here @wkurniawan07 - looks good. 

Wondering how I go about testing this locally? i.e Is there anything specifically I need to change to build and run the web server

My comments are below:

File: src/main/java/teammates/ui/webapi/OriginCheckFilter.java (hunk 4 above)
Comment (wkurniawan07) [reply src/main/java/teammates/ui/webapi/OriginCheckFilter.java:79]: > Since the filter will never be called for those worker/ or auto/ App Engine Queue tasks we have.

No longer true after the change introduced in this PR.

As for why the origin check filter needs to be extended to them, it's because our APIs are wide open to public (i.e. not restricted by firewalls/IPs) and we need to grant the same security as we do to the more normal (for the lack of better word) APIs in `webapi/`.

File: src/main/java/teammates/ui/webapi/RequestTraceFilter.java (hunk 3 above)
Comment (wkurniawan07) [reply src/main/java/teammates/ui/webapi/RequestTraceFilter.java:52]: I don't have data to back up but I don't believe tearing down will take that long.

File: src/main/java/teammates/common/util/Logger.java (hunk 1 above)
Comment (wkurniawan07) [reply src/main/java/teammates/common/util/Logger.java:112]: StackWalker is a Java 9 feature.

File: src/main/java/teammates/common/util/Logger.java (hunk 2 above)
Comment (wkurniawan07) [reply src/main/java/teammates/common/util/Logger.java:122]: Any idea to figure out "when we are at the bottom of the stack"?

File: src/main/java/teammates/common/util/Logger.java (hunk 1 above)
Comment (rrtheonlyone) [reply src/main/java/teammates/common/util/Logger.java:112]: Ah yes - well something to keep in view after we upgrade the java version

File: src/main/java/teammates/common/util/Logger.java (hunk 2 above)
Comment (rrtheonlyone) [reply src/main/java/teammates/common/util/Logger.java:122]: An alternative might be to just throw an exception since it's not something that should be expected e.g. this function should not be called when there is nothing on the stack

Comment (wkurniawan07) [reply src/main/java/teammates/common/util/Logger.java:122]: I don't really like this class interrupting program flow. This is an internal system logger; the business logic of the system performs just fine even if this class does not exist.
If an exception is to be thrown here only to be caught and handled by the same class, it is effectively the same as the current implementation.

Comment (rrtheonlyone) [reply src/main/java/teammates/common/util/Logger.java:122]: Yes that is a fair point

**Review by rrtheonlyone — APPROVED** [review]
LGTM :+1:

**Review by jianhandev — APPROVED** [review]
LGTM
